### PR TITLE
Token Management API

### PR DESCRIPTION
**Pull request by royalchan-okta**
## Description:
- Token Management APIs for : User-Client, Authorization Server-Client, Application

### Resolves:

* [OKTA-145525](https://oktainc.atlassian.net/browse/OKTA-145525)

@mystiberry-okta

### CODE REVIEW

File: _source/_docs/api/resources/authorization-servers.md [hunk 1]
```diff
@@ -1522,4 +1522,288 @@ This object defines a [JSON Web Key Set](https://tools.ietf.org/html/rfc7517) fo
...
+
+### List Client Resources for an Authorization Server
+{:.api .api-operation}
+
+{% api_lifecycle ea %}
+
+{% api_operation get /api/v1/authorizationServers/${authorizationServerId}/clients %}
+
+Lists all client resources for which the specified authorization server has tokens.
+
```
Comment (mystiberry-okta) [inline _source/_docs/api/resources/authorization-servers.md]: These intros aren't complete sentences so we leave the period off.

File: _source/_docs/api/resources/authorization-servers.md [hunk 2]
```diff
@@ -1522,4 +1522,288 @@ This object defines a [JSON Web Key Set](https://tools.ietf.org/html/rfc7517) fo
...
+~~~
+
+
+#### Get Token for Authorization Server and Client
+{:.api .api-operation}
+
+{% api_operation get /api/v1/authorizationServers/${authorizationServerId}/clients/${clientId}/tokens/${tokenId} %}
+
+Get a token for the specified authorization server and client.
+
```
Comment (mystiberry-okta) [inline _source/_docs/api/resources/authorization-servers.md]: to match other intro fragments: "Get**s** a token for the specified authorization server and client" (no period)

**Comment by mystiberry-okta** [comment]
See my nits plus two questions:

1. Looks like this is going out next week, so I'm changing the base to `weekly`. We won't merge into weekly until the feature is merged. If that doesn't get rid of travis test issues, I'll ping you.
2. Does this manage ALL tokens (even API Token)? If not, should we specify which tokens aren't included?

**Comment by mystiberry-okta** [comment]
You just need to fix one broken link: 

linking to internal hash #revoke-all-tokens-for-authorization-server-and-client that does not exist (line 2560)
     <a href="#revoke-all-tokens-for-authorization-server-and-client">Revoke Tokens for Authorization Server and Client</a>

**Comment by royalchan-okta** [comment]
@mystiberry-okta this is just OAuth2 Tokens, so i'll make it clear everywhere

File: _source/_docs/api/resources/apps.md [hunk 3]
```diff
@@ -4416,6 +4416,244 @@ curl -v -X GET \
...
+Lists all tokens for the application
+
+#### Request Parameters
+{:.api .api-request .api-request-params}
+
+| Parameter     | Description                                                                                  | Param Type | DataType | Required | Default |
+|:--------------|:---------------------------------------------------------------------------------------------|:-----------|:---------|:---------|:--------|
+| applicationId | ID of the application                                                                        | URL        | String   | TRUE     |         |
+| expand        | Valid value: `scope`. If specified, scope details are included in the `_embedded` attribute. | Query      | String   | FALSE    |         |
+| limit         | The maximum number of tokens to return                                                       | Query      | Number   | FALSE    | 20      |
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/apps.md]: Should we add a section below to explain that the max limit is 200?

File: _source/_docs/api/resources/apps.md [hunk 4]
```diff
@@ -4416,6 +4416,244 @@ curl -v -X GET \
...
+~~~sh
+HTTP/1.1 204 No Content
+~~~
+
+### Revoke Token for Applications
+{:.api .api-operation}
+
+{% api_lifecycle ea %}
+
+{% api_operation delete /api/v1/apps/${applicationId}/tokens %}
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/apps.md]: tokens -> tokens/${tokenId}

File: _source/_docs/api/resources/authorization-servers.md [hunk 5]
```diff
@@ -1522,4 +1522,288 @@ This object defines a [JSON Web Key Set](https://tools.ietf.org/html/rfc7517) fo
...
+| after                 | Specifies the pagination cursor for the next page of grants                                  | Query      | String   | FALSE    |         |
+
+##### Request Example
+{:.api .api-request .api-request-example}
+
+~~~sh
+curl -v -X GET \
+-H "Accept: application/json" \
+-H "Authorization: SSWS ${api_token}" \
+"https://{yourOktaDomain}.com/api/v1/authorizationServers/ausnsopoM6vBRB3PD0g3/clients/0oabskvc6442nkvQO0h7/tokens"
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/authorization-servers.md]: Same here, I think we should include the expand param to show how it looks with the _embedded attribute.

File: _source/_docs/api/resources/authorization-servers.md [hunk 6]
```diff
@@ -1522,4 +1522,288 @@ This object defines a [JSON Web Key Set](https://tools.ietf.org/html/rfc7517) fo
...
+    }
+  }
+]
+~~~
+
+
+#### Get Token for Authorization Server and Client
+{:.api .api-operation}
+
+{% api_operation get /api/v1/authorizationServers/${authorizationServerId}/clients/${clientId}/tokens/${tokenId} %}
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/authorization-servers.md]: image
This isn't fortunate, any way we can get this to look better @mystiberry-okta ? It is not related to my browser width, as the page is a fixed size.

File: _source/_docs/api/resources/authorization-servers.md [hunk 7]
```diff
@@ -1522,4 +1522,288 @@ This object defines a [JSON Web Key Set](https://tools.ietf.org/html/rfc7517) fo
...
+~~~sh
+HTTP/1.1 204 No Content
+~~~
+
+### Revoke Token for Authorization Server and Client
+{:.api .api-operation}
+
+{% api_lifecycle ea %}
+
+{% api_operation delete /api/v1/authorizationServers/${authorizationServerId}/clients/${clientId}/tokens/${tokenId} %}
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/authorization-servers.md]: image

Same issue here.

File: _source/_docs/api/resources/users.md [hunk 8]
```diff
@@ -3058,27 +3057,267 @@ curl -v -X DELETE \
...
+
+#### Request Example
+{:.api .api-request .api-request-example}
+
+~~~sh
+curl -v -X GET \
+-H "Accept: application/json" \
+-H "Content-Type: application/json" \
+-H "Authorization: SSWS ${api_token}" \
+"https://{yourOktaDomain}.com/api/v1/users/00u5t60iloOHN9pBi0h7/clients/0oabskvc6442nkvQO0h7/tokens"
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/users.md]: Same here, would be nice with expand.

File: _source/_docs/api/resources/users.md [hunk 9]
```diff
@@ -3058,27 +3057,267 @@ curl -v -X DELETE \
...
+  "status": "ACTIVE",
+  "created": "2018-03-09T03:18:06.000Z",
+  "lastUpdated": "2018-03-09T03:18:06.000Z",
+  "expiresAt": "2018-03-16T03:18:06.000Z",
+  "issuerId": "ausain6z9zIedDCxB0h7",
+  "clientId": "0oabskvc6442nkvQO0h7",
+  "userId": "00u5t60iloOHN9pBi0h7",
+  "scopes": [
+    "offline_access",
+    "car:drive"
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/users.md]: Do we have any consistency with what we use for custom scopes in the doc?

File: _source/_docs/api/resources/users.md [hunk 10]
```diff
@@ -3058,27 +3057,267 @@ curl -v -X DELETE \
...
+HTTP/1.1 204 No Content
+~~~
+
+### Revoke Token for User and Client
+{:.api .api-operation}
 
 {% api_lifecycle ea %}
 
-Each grant references a user and a client.
+{% api_operation delete /api/v1/users/${userId}/clients/${clientId}/tokens %}
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/users.md]: ```tokens-> tokens/${tokenId}```

File: _source/_docs/api/resources/users.md (hunk 9 above)
Comment (royalchan-okta) [reply _source/_docs/api/resources/users.md]: i just copied `car:drive` this from another example, `offline_access` is required

File: _source/_docs/api/resources/users.md [hunk 11]
```diff
@@ -3105,11 +3344,15 @@ curl -v -X GET \
             "grants": {
                 "href": "https://{yourOktaDomain}.com/api/v1/users/00u5t60iloOHN9pBi0h7/clients/0oabskvc6442nkvQO0h7/grants"
             }
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/users.md]: Missing ,

File: _source/_docs/api/resources/apps.md (hunk 3 above)
Comment (royalchan-okta) [reply _source/_docs/api/resources/apps.md]: yeah, we are also missing cursor info for the grant apis

Comment (royalchan-okta) [reply _source/_docs/api/resources/apps.md]: image

File: _source/_docs/api/resources/apps.md [hunk 12]
```diff
@@ -4416,6 +4416,244 @@ curl -v -X GET \
...
+
+#### Request Example
+{:.api .api-request .api-request-example}
+
+~~~sh
+curl -v -X GET \
+-H "Accept: application/json" \
+-H "Content-Type: application/json" \
+-H "Authorization: SSWS ${api_token}" \
+"https://{yourOktaDomain}.com/api/v1/apps/0oabskvc6442nkvQO0h7/tokens"
```
[the remark this answers is not in the document]
Comment (royalchan-okta) [reply _source/_docs/api/resources/apps.md]: probably better to do both requests, but thats a lot fo example json

File: _source/_docs/api/resources/authorization-servers.md (hunk 7 above)
Comment (royalchan-okta) [reply _source/_docs/api/resources/authorization-servers.md]: hm this is kind of hard to fix, unless we shorten something like `authorizationServerId` -> `asId`

File: _source/_docs/api/resources/users.md (hunk 8 above)
Comment (royalchan-okta) [reply _source/_docs/api/resources/users.md]: see other comment

File: _source/_docs/api/resources/authorization-servers.md (hunk 5 above)
Comment (royalchan-okta) [reply _source/_docs/api/resources/authorization-servers.md]: see other comment

File: _source/_docs/api/resources/authorization-servers.md (hunk 7 above)
Comment (mystiberry-okta) [reply _source/_docs/api/resources/authorization-servers.md]: @jakubvul-okta It appears we have a length limit. Should we go with authServerId if that makes it short enough? or is there some other fix?

Comment (royalchan-okta) [reply _source/_docs/api/resources/authorization-servers.md]: chatted with Mysti, shorting this to `authServerId` for now

File: _source/_docs/api/resources/apps.md (hunk 3 above)
Comment (larsjohansen-okta) [reply _source/_docs/api/resources/apps.md]: Looks ok to me, for styling preference I'd refer to @mystiberry-okta to approve

File: _source/_docs/api/resources/apps.md [hunk 13]
```diff
@@ -4416,6 +4416,272 @@ curl -v -X GET \
...
+|:--------------|:---------------------------------------------------------------------------------------------|:-----------|:---------|:---------|:--------|
+| applicationId | ID of the application                                                                        | URL        | String   | TRUE     |         |
+| tokenId       | ID of the token                                                                              | URL        | String   | TRUE     |         |
+| expand        | Valid value: `scope`. If specified, scope details are included in the `_embedded` attribute. | Query      | String   | FALSE    |         |
+| limit         | The maximum number of grants to return                                                       | Query      | Number   | FALSE    | 20      |
+| after         | Specifies the pagination cursor for the next page of grants                                  | Query      | String   | FALSE    |         |
+
+> Note: The after cursor should treated as an opaque value and obtained through [the next link relation](/docs/api/getting_started/design_principles#pagination).
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/apps.md]: I think we should include the limit max value here as well?

Comment (royalchan-okta) [reply _source/_docs/api/resources/apps.md]: ah good catch, i *shouldn't* have the cursor params here

File: _source/_docs/api/resources/users.md [hunk 14]
```diff
@@ -3058,27 +3075,295 @@ curl -v -X DELETE \
...
+| userId    | ID of the user for whom you are fetching tokens                                              | URL        | String   | TRUE     |         |
+| clientId  | ID of the client                                                                             | URL        | String   | TRUE     |         |
+| tokenId   | ID of the token                                                                             | URL        | String   | TRUE     |         |
+| expand    | Valid value: `scope`. If specified, scope details are included in the `_embedded` attribute. | Query      | String   | FALSE    |         |
+| limit     | The maximum number of grants to return                                                       | Query      | Number   | FALSE    | 20      |
+| after     | Specifies the pagination cursor for the next page of grants                                  | Query      | String   | FALSE    |         |
+
+> Note: The after cursor should treated as an opaque value and obtained through [the next link relation](/docs/api/getting_started/design_principles#pagination).
```
Comment (larsjohansen-okta) [inline _source/_docs/api/resources/users.md]: Should we include the limit max value here?

File: _source/_docs/api/resources/apps.md (hunk 3 above)
Comment (royalchan-okta) [reply _source/_docs/api/resources/apps.md]: I actually made it match other places: 
image

**Pull request merged**